---
int8 ret
string retStr